[WIP] job-info module: get more info beyond eventlog

chu11 opened the pull request:
I began with the simple goal of supporting getting the job's eventlog and jobspec before moving onto a bulk read feature.  If I can get jobspec working nicely/cleanly, then I think adding any other job fields will be easy.

But hit some issues.  Here's the gotchas and thoughts.

- Initially, my plan was to do future composites like this:

```
  f1 = flux_kvs_lookup ("eventlog path");
  f2 = flux_future_and_then (f1, eventlog_guest_continuation);
  f3 = flux_future_and_then (f2, jobspec_continuation);
```

Where `eventlog_guest_continuation()` would do the guest access check, then do another lookup leading to `jobspec_continuation()`.  But the gotcha is that `eventlog_guest_continuation()` may not find the eventlog in the `active` path, but in the `inactive` path.  If it's in the latter, I get an ENOENT, gotta do a new lookup, and now the future chain is broken.

I wasn't sure if there was a way to deal with this situation, other than rebuild the whole chain again from scratch?

- how would we want to provide this feature to users?  Via additional `flux job <subcommand>` commands?  Or perhaps a new `flux job info [ eventlog | jobspec | R]` etc new command?

- Specifically for `flux job jobspec`, do we want to support any nicer outputs for the end user?

grondo commented:
> I wasn't sure if there was a way to deal with this situation, other than rebuild the whole chain again from scratch?

Can you use `flux_future_or_then(3)`? I'm curious because this situation is what it is meant for, however, it hasn't been used IRL so I'm not sure if it has enough functionality.

chu11 commented:
> Can you use flux_future_or_then(3)? I'm curious because this situation is what it is meant for, however, it hasn't been used IRL so I'm not sure if it has enough functionality.

Are you suggesting something like:

```
  f1 = flux_kvs_lookup ("eventlog path");
  f2-active = flux_future_and_then (f1, eventlog_guest_continuation);
  f2-inactive = flux_future_or_then (f1, eventlog_inactive_continuation);
  f3-active = flux_future_and_then (f2-active, jobspec_continuation);
```

where in `eventlog_inactive_continuation()` I could rebuild the chain again?

chu11 commented:
FWIW, in my mind, I imagining something like this.  I do this:

```
  f1 = flux_kvs_lookup ("eventlog path");
  f2 = flux_future_and_then (f1, eventlog_guest_continuation);
  f3 = flux_future_and_then (f2, jobspec_continuation);
```

in `eventlog_guest_continuation()`

```
if (flux_kvs_lookup_get (f) < 0) {
   if (errno == ENOENT && <some flag says I tried active path>) {
       flux_future_t *fnew = flux_kvs_lookup (<inactive eventlog path>);
       flux_future_and_then (fnew, eventlog_guest_continuation)
       flux_future_and_then_override (fnew, steal_flux_future_and_thens (f))
   }
   /* other error handling */
}
```

So basically, whatever future "and thens" were attached to `f`, steal then, and put them onto `fnew`.  The reason is because I imagine during large bulk loads in the future we'll have:

```
  f1 = flux_kvs_lookup ("eventlog path");
  f2 = flux_future_and_then (f1, eventlog_guest_continuation);
  f3 = flux_future_and_then (f2, jobspec_continuation);
  f4 = flux_future_and_then (f3, R_continuation);
  f5 = flux_future_and_then (f4, FOO_continuation);
```

and in any of the continuations above, the data could be moved from `active` to `inactive` in any of the continuations.  So instead of tracking which continuations have been completed and which haven't, one could just "steal" the remaining ones and continue along.

This probably isn't doable for the moment.  But wasn't sure how best to handle this in the mean time.

grondo commented:
I *think* `flux_future_or_then(3)` should return the same "next" future as `flux_future_and_then(3)` when called on the same "prev" future. Therefore, as in the documentation:

> This function is useful for recovery or other error handling
> (other than the default behavior of propagating an error down the chain
> to the final result). The flux_future_or_then() callback offers a
> chance to successfully fulfill the "next" future in the chain, even
> when the "previous" future was fulfilled with an error.

Try something like:

```C
f1 = flux_kvs_lookup ("eventlog path");
f2 = flux_future_and_then (f1, eventlog_guest_continuation);
f3 = flux_future_or_then (f1, eventlog_inactive_continuation);
assert (f2 == f3); 
f4 = flux_future_and_then (f3, jobspec_continuation);
```

Either the `and_then` or the `or_then` can call `flux_future_continue()` to continue the chain (without error). The `eventlog_inactive_continuation` should of course check to ensure the expected error was received..

If that doesn't work then we've got some fixin' to do.

grondo commented:
For fetching multiple keys, wouldn't you want to do those in parallel instead of sequentially? Maybe you want to create the composite future as a series of `flux_kvs_lookup(), flux_future_or_then` requests, all going in parallel as part of a `wait_all` composite?

Hm, although do you *first* have to process the eventlog to determine the guest user's id?

grondo commented:
BTW, you could set the same callback for both `flux_future_and_then()` and `flux_future_or_then()` -- we could add a convenience function to do that `flux_future_and_or_then`?

Also, I found in the job-exec module it wasn't easy, or I failed, to create nice clean composite futures to do complex tasks. What I ended up with was a hybrid approach. I defined complex operations *within* continuation callbacks that were then chained together. This was a bit easier to wrap my head around, though there may be a better way.

For example, the "cleanup" for a job that has completed execution requires the following steps:

 1. emit a cleanup event
 2. then: start all "cleanup" tasks in parallel
 3. then: emit a cleanup.finish event
 4. then: emit `done` to exec.eventlog
 5. then: copy namespace to primary
 6. then: remove namespace
 7. then: send release event to job-manager

1,2,3 are implemented as a chained future (`jobinfo_start_cleanup`), where 2 is a `wait_all` composite future (`jobinfo_cleanup`), 4,5,6 are also a implemented as a chained future (`namespace_move`), and finally the release (`jobinfo_release`) is an open loop final response to job-manager, after which the job info object is destroyed.

So the general flow is something like

```
f = jobinfo_start_cleanup (job);
if (job->has_namespace)
    fnext = flux_future_and_then (f, namespace_move, job);
flux_future_then (fnext, -1., jobinfo_release, job);
```


Hope this helps!

chu11 commented:
>  Hope this helps!

I think I may have not groked some of the subtleties of the composite/chaining stuff.  I think I need to re-look through it.

> although do you first have to process the eventlog to determine the guest user's id?

Ahh, I didn't think about that.  I suppose I could do the guest check in parallel with the lookups, and just make sure that it and all lookups have been satisfied before returning to the user.

grondo commented:
> I think I may have not groked some of the subtleties of the composite/chaining stuff. I think I need to re-look through it.

The implementation is simple if you look at the code, but the interface is a little tricky for some reason in C (I think it is confusing what those intermediate "next" futures really are)

> Ahh, I didn't think about that. I suppose I could do the guest check in parallel with the lookups, and just make sure that it and all lookups have been satisfied before returning to the user.

Or you could first fetch eventlog and do access check, then if that is successful start the other work in parallel and return to user once all items have been fetched (avoiding unnecessary lookups to the kvs)

chu11 commented:
> Or you could first fetch eventlog and do access check, 

This is probably the best path, because there's a special case.  If the user only requested the eventlog, no additional KVS lookups have to occur.

> then if that is successful start the other work in parallel and return to user once all items have been fetched (avoiding unnecessary lookups to the kvs)

There's one subtlety here.  The guest access check could lookup the eventlog in `job.active`, but by the time the lookup of other job information starts, the information may have been moved to `job.inactive`.  But I think this situation will be rare enough, that in the event it happens, we can live with the extra KVS lookups that occur.

garlick commented:
> Or you could first fetch eventlog and do access check, then if that is successful start the other work in parallel and return to user once all items have been fetched (avoiding unnecessary lookups to the kvs)

One idea I think we bounced around was caching the id:owner mapping for a period of time so that repeated requests for attributes from the same job don't have to cause the same eventlog to be fetched again and again.  I guess actually that's just an optimization and can be added later if it seems useful.

garlick commented:
I'm not sure this is a great idea, but if the job-info module subscribed to the `job-state` events from pr #2109 when it is first loaded, it could build a hash of all jobids in the system, and track when they transition to inactive state (to know where to look once we start moving them).  If we add an event from the job-ingest manager that includes the userid, you could just use that for authentication...

Big downside: tracking all that in memory.  Although just a hash of jobid => (userid, state) might not be too big of a deal... (12 bytes per job, so 1M jobs takes 12MB, plus hash table overhead)

grondo commented:
I really like that idea as a use case for the job state transition events.

Are there some small races though? e.g. could a request come in for a job before the first state transition event is received by the module, or could a job transition to inactive just after the hash lookup reads the previous state?  Even so, this just means that the module always has to fall back to the inactive lookup, and jobs not hashed still have to go through an active/inactive lookup?

garlick commented:
Yes that's true.  If you submit a job and immediately try to fetch info on it, the job-info module might not have received the notification yet.  Yeah, maybe not the best fit....

grondo commented:
Is there some other approach we're missing to simplify the active/inactive transition? IIRC, the main reason for the `jobs.active` directory is to limit the scanning required for a job-manager restart. Could we also provide a `jobs.all.` directory with links to the jobs in active on creation, and updated to `jobs.inactive.` in the same commit when the job is moved?

(Alternately I guess, `jobs.active` could become links, which could just be removed when the jobs are no longer active.)

garlick commented:
Brain storming a bit:  the job manager knows of all the active jobs.  Is there anything it can do to make life easier here?  For example, it could keep all inactive jobs in a hash with userid and KVS snapshot reference, and answer (for `job-info`): 
* is job ID valid?
* who owns job ID?
* is job ID active or inactive?
* (maybe) what is final KVS snapshot for inactive KVS job ID ?

The other random thought is at one point we had discussed dumping inactive jobs into a database of some sort.  Would taking a step in that direction help here?

garlick commented:
Ugh, except if active, there is still a race where job could transition to inactive.  Maybe it would need to track a snapshot ref (or at least a KVS root ref) for all jobs in order to be helpful here?

chu11 commented:
> Ugh, except if active, there is still a race where job could transition to inactive. Maybe it would need to track a snapshot ref (or at least a KVS root ref) for all jobs in order to be helpful here?

Yeah, that was my immediate first thought.  In my two-phase lookup (userlog for access, then lookup actual data), each has to check for active -> inactive transition b/c it can happen at any point.

Food for thought, you originally considered having the "job info proxy" in job manager.  Is it worthwhile considering moving it back in there if the job-manager simply has that much info that would be useful to the "job info proxy"?

grondo commented:
The nice thing about the dual-lookup is that it is race-free and will just kind of work. Once you've encoded the dual lookup into a chain it isn't that much code, so is it worth adding complexity to synchronize with job-manager or return snapshots to work around it? I'm honestly asking, because I'm not sure of the real impact.

Also, do we really not want to reexamine the active-inactive kvs directory layout. If the only reason to have a separate directory is so the job-manager can restart (something I don't think we plan to do at this point), is it really worth the other side-effects we're discussing here?

garlick commented:
Sorry, agreed, let's reexamine that design.

If we want to just assume in this PR that nothing moves to inactive, I'm fine with that for now and we can talk through the longer term plan when we're all together again?  The move to inactive is not implemented in the current code base so any tests would need to be somewhat contrived here anyway.

The symlink idea could be employed if we find we do need this down the road.

We should put crazy on the table too, in case anybody comes up with a unique and cool idea here!  I don't have one at the moment, I'm sorry to say.

chu11 commented:
just pushed try #2

I used composite futures for the parallel KVS lookup of all info requested from the caller.

But one ugly-ish part is the building up of the response, b/c it is variable depending on what users input.  It's not something (I believe) can be done with any of the pack functions? it's like this pseudo-code:

```
o = json_object ();
foreach info:
   str = json_string (data);
   json_object_set (o, name, str);
   json_decref (str);
response_data = json_dumps (o);
flux_respond (response_data);
```

I created a new `flux job info` command where users can specify any number of `--eventlog | --jobspec | --J | --R`.  No output formatting, just dumps raw to stdout for now.  Possibly can replace `flux job
eventlog` in future?  But it's sort of weird to have options like `--context-format` when it only applies to one of the options?  Maybe it should just be a testing tool.  But can deal with that later.

One aside, I wasn't 100% sure when jobspec, J, and R are available in the KVS.  I believe jobspec & J are immediately available after ingest, and R is available after `alloc`?

grondo commented:
> One aside, I wasn't 100% sure when jobspec, J, and R are available in the KVS. I believe jobspec & J are immediately available after ingest, and R is available after alloc?

Yes, that  is correct. A scheduler commits `R` to the kvs first then responds to the job-manager's `alloc` request, which results in an `alloc` event posted to the eventlog.

> But it's sort of weird to have options like --context-format when it only applies to one of the options? Maybe it should just be a testing tool. But can deal with that later.

Yeah, it seems fine to me for now. Likely the real use cases will be using the rpc directly.  I actually like that there's a plumbing tool that dumps the raw values, could be useful in both testing and quick shell scripts. (I agree though that `flux job eventlog` deserves its own command, since it reformats the eventlog which will soon be json.)

For the other keys, I wonder if something like `flux job getinfo [OPTIONS] ITEM  [ITEM]...` might work better than requiring an option per item, then the policy about what get be fetched can be in the module, and new items can be added without requiring modification of the utility (not that I can think of what else might be fetched).

chu11 commented:
> For the other keys, I wonder if something like flux job getinfo [OPTIONS] ITEM [ITEM]... might work better than requiring an option per item, then the policy about what get be fetched can be in the module, and new items can be added without requiring modification of the utility (not that I can think of what else might be fetched).

You mean like `flux job getinfo eventlog jobspec`?  With the current implementation, I set flags (e.g. `FLUX_JOB_INFO_JOBPSEC`) that informs the module what to fetch.  So it would still require modification of the utility.

Instead of flags, perhaps we could send the strings directly over to the module?  Then whatever strings are sent over to the module are directly looked up?  That way additional job information can be added in the future by the job manager, and `job-info` and `flux job getinfo` don't have to be modified at all.

garlick commented:
> Instead of flags, perhaps we could send the strings directly over to the module?

Like!

chu11 commented:
I redid the module so that instead of passing a flag to indicate desire to read eventlog, jobspec, etc., you just pass in an array of job keys.  i.e. `eventlog`, `R`, etc.  If all of the data you request is available, it returns all of it, otherwise you get ENOENT or an appropriate error.

chu11 commented:
hmmm, one builder hit

```
ERROR: t2205-job-info-security.t - missing test plan
ERROR: t2205-job-info-security.t - exited with status 143 (terminated by signal 15?)
```

```
2019-04-01T01:46:42.950034Z broker.err[0]: runlevel 2 timeout, sending SIGTERM
2019-04-01T01:46:42.950575Z broker.err[0]: Run level 2 Unknown signal 143 (rc=143) 300.0s
```

I guess it took too long?  Was it just unluckily slow or did it hit a hang?  Lets try again.

garlick commented:
Playing around with this a bit and it seems solid so far!  I'll work on a more detailed review tomorrow.  Offhand I noticed this residue in lookup.c from the previous approach:
```
/* maximum bitmask that can exist */
#define FLUX_JOB_INFO_BITMASK_MAX FLUX_JOB_INFO_R
```
and a rolemask bit check that probably needs the double parens.  (Just minor but thought I'd mention)

chu11 commented:
@garlick doh!  I fixed up those nits.

chu11 commented:
diff target is bad, I can eek out a bit more by adding an "inactive" guest user access check in the eventlog (will be in the next push), but it's probably going to stay bad overall given error paths.

garlick commented:
Peeking at `lookup.c` in a bit more detail - would it simplify the code a bit and improve latency for the common (success) case if the eventlog was pushed into the same composite future as other info keys and the access check was deferred until the composite is fulfilled?

garlick commented:
It might be useful to add this to the sharness valgrind workload to be sure an info fetch doesn't leak memory in the future (looks good now):
```diff
diff --git a/t/valgrind/workload.d/job b/t/valgrind/workload.d/job
index 5c99b4444..07d038d01 100755
--- a/t/valgrind/workload.d/job
+++ b/t/valgrind/workload.d/job
@@ -15,4 +15,7 @@ flux job wait-event ${id} start
 flux job cancel ${id}
 flux job wait-event ${id} clean

+# Test info fetch
+flux job info ${id} eventlog jobspec R
+
 flux job drain
```

chu11 commented:
> Peeking at lookup.c in a bit more detail - would it simplify the code a bit and improve latency for the common (success) case if the eventlog was pushed into the same composite future as other info keys and the access check was deferred until the composite is fulfilled?

I went back and forth on this, but now that the code looks the way it does, I think it's worth considering putting everything into one composite future and one continuation.  Let me refactor and see if that works out.

chu11 commented:
I forgot to mention.  In order to be able to read `R`, I of course had to load more modules.  I have absolutely no idea why I have to hack in the `hwloc_fake_config='{"0-3":{"Core":2,"cpuset":"0-1"}}'` below.  I just copied it from elsewhere.  If I didn't, I got hangs / errors.

```
+hwloc_fake_config='{"0-3":{"Core":2,"cpuset":"0-1"}}'
+
+test_expect_success 'load job-exec,sched-simple modules' '
+        #  Add fake by_rank configuration to kvs:
+        flux kvs put resource.hwloc.by_rank="$hwloc_fake_config" &&
+        flux module load -r 0 sched-simple &&
+        flux module load -r 0 job-exec
+'
```

garlick commented:
OK I'll hold off on more review until you get that in.  Not sure if you've had a chance to look at #2109 but that one's ready to go in IMHO.  No @grondo this week.

garlick commented:
> I forgot to mention. In order to be able to read R, I of course had to load more modules. 

The scheduler relies on hwloc being properly initialized in rc1.  You might just want to switch to the "default" rc1/rc3 by dropping the "personality" argument from `test_under_flux` if there is no need to customize the set of modules loaded.

chu11 commented:
> In order to be able to read R

Oh wait, now that I'm not using flags, I actually don't have to test reading R.

chu11 commented:
ok, re-pushed.  Changed and added a bunch of tests.  Since I'm no longer using "flags" in the `job-info` module to determine what data a user wants to read, I don't need to test retrieving `R` anymore.  Testing for key-names is all that is needed.  Also added some common functions.

Last commit refactors all info & guest access lookups into a single composite.  Getting rid of the extra latency of the original "guest access" lookup wasn't as clean as I would have hoped,  But I think it's definitely a net win.

The main logic trickiness is:

- did the user request "eventlog"? if the user did not request "eventlog", have to add it to the list of keys to retrieve.
- in the continuation, was "eventlog" requested by the user or did we add it?  If the later, we don't want to return eventlog data to the caller

garlick commented on src/cmd/flux-job.c:
```diff
@@ -857,27 +857,38 @@ int cmd_wait_event (optparse_t *p, int argc, char **argv)
         if (errno == ENOENT) {
             flux_future_destroy (f);
-            log_msg_exit ("job %lu.%s not found", ctx->id, ctx->key);
+            log_msg_exit ("job %lu id or key not found", id);
```
Might be nice to return a text string from the service so that one knows whether the job or the key is wrong.

chu11 replied:
part of the issue is that we don't know.  It's simply an ENOENT from the KVS.  It could be b/c the jobid is missing or the key is missing.

Hmmm.  Since the infrastructure does everything in parallel now, we could add an additional lookup on just `jobs.active.jobid` and that can give us info for an error return string?

garlick replied:
Perhaps if the eventlog is missing, we know it's a bad jobid since every valid jobid has an eventlog from the very start?  If eventlog is valid and key isn't then the key is likely missing.

chu11 replied:
Hmmm. That would require us to lookup the eventlog even when we don't need to (i.e. we already checked that the instance owner sent the rpc).  

If we looked up the eventlog everytime, i guess it wouldn't be the end of the world.  Would we want to add it just for the nicer error msg?

garlick replied:
Ooops I forgot that we don't read it every time.  Well, for now maybe not worth it.

However it may still be useful to return a textual error message.  For example, if lookup of a particular key returns ENOENT we could at least say which one of the set caused the failure.

garlick replied:
Or maybe it is worth it.  I was thinking of the case where a queue listing tool needs to fetch info for every job, where the extra kvs lookups might add up to something substantial.  But I suppose in that case, most likely the tool is going to fetch everything anyway.  In the single job case, fetching it in parallel with other things won't be a big deal.

What is your opinion?

chu11 replied:
that makes sense, lets fetch it everytime.

chu11 replied:
This ended up trickier than I had hoped.  Because `active` to `inactive` transition can happen at anytime, I believe it's possible to get success on one lookup and ENOENT on another lookup done in parallel.

So what I've been doing prior to this is "switch" to inactive whenever I get an ENOENT on any key that is missing in the `active` path.

Assume a job is still active.  What if the user inputs `flux job info 12345 eventlog jobspec badkey`.  The `job-info` module notices ENOENT on `badkey`, switches to `inactive` and notices `eventlog` is missing in `inactive`.  Now I need to manage that I found `eventlog` in the active path prior.  Also I have to iterate through all keys to check if `eventlog` was found or not.

I think looking up a `jobs.active.jobid` would be the right path now.  We can also drop this issue until later.

garlick replied:
It seems like it's the same problem whether we look up `jobs.active.jobid` or `jobs.active.jobid.eventlog`.  Both would transition in the same way.

This inactive archival plan needs a rethink (as discussed last week with @grondo), so since we haven't implemented it yet anyway, maybe just remove the active/inactive logic from this part of the code now.

Sorry this design proved to be such a pain.

chu11 replied:
The PR is getting a bit long, so I'm going to start splitting it up.  How about we save inactive removal for another PR.

garlick replied:
Fine by me.

garlick commented on src/modules/job-info/lookup.c:
```diff
@@ -102,58 +100,137 @@ static flux_future_t *lookup_key (struct lookup_ctx *l, const char *key,
+        }
+    }
+
+    if (flux_future_then (fall, -1, c, l) < 0) {
```
If this fails, `f` is freed but it was already added to `fall` so double free?

Note that `flux_future_destroy()` doesn't clobber errno so it's not strictly necessary to save/restore (which is convenient to know if you have to do the free in the `flux_future_push()` error case rather than in `error:`.

garlick commented on src/modules/job-info/lookup.c:
```diff
@@ -234,64 +259,45 @@ static void info_lookup_continuation (flux_future_t *fall, void *arg)
+/* If keys array doesn't contain eventlog, add it
+ * and update flag indicating we added it.
+ */
+static int check_keys_for_eventlog (struct lookup_ctx *l)
```
Maybe it would simplify the code if the eventlog was not added to the json array?  Just check if it's there and if not, explicitly look it up and insert a future named `eventlog` into the composite.  Then in the continuation, first thing (if `!allow`), fetch that future from the composite by name, perform the access check, and fail right there if denied.

Then instead of iterating over the futures in the composite, iterate over the names in the json array when constructing the response.

garlick replied:
That one was meant as an idea that you might consider but choose to veto as more trouble than it's worth.

chu11 replied:
after your comment above about the error string, it occurred to me that I could do something like this.  Let me see how it plays out, might be cleaner overall.

chu11 commented on src/modules/job-info/lookup.c:
```diff
@@ -329,9 +335,11 @@ void lookup_cb (flux_t *h, flux_msg_handler_t *mh,
         /* must lookup eventlog first to do access check */
-        if (!(f = lookup_key (l, "eventlog",
-                              eventlog_lookup_continuation)))
+        if (!(f = lookup_keys (l, info_lookup_continuation)))
```
ahh, now that the code falls that way we can definitely clean that up.  Thanks for the catch.

garlick commented:
One more suggestion: drop the libjob API for the eventlog fetch, since it only has the one user in flux-job.c, and the other fetchable items don't have API representation.

chu11 commented:
just re-pushed fixing up suggestions from above.  I think the code cleaned up a lot more nicely than I had expected.

chu11 commented:
closing, as this PR is being split off into multiple smaller ones.
